File: src/web_optimise/_commands.py
"""ffmpeg command builders for web-optimised encoding."""

from typing import TYPE_CHECKING

from web_optimise._constants import B_FRAMES
from web_optimise._constants import CRF
from web_optimise._constants import KEYFRAME_INTERVAL_SECS
from web_optimise._constants import LEVEL
from web_optimise._constants import PRESET
from web_optimise._constants import PROFILE
from web_optimise._constants import X264_PARAMS

if TYPE_CHECKING:
    from pathlib import Path

    from web_optimise._types import FileInfo


def build_ffmpeg_cmd(
    info: FileInfo,
    output_path: Path,
    /,
) -> tuple[str, ...]:
    """
    Build the ffmpeg command for web-optimised encoding.

    Produce H.264 High profile, 2-second keyframes,
    stream-copied audio, faststart, stripped metadata.
    """
    fps = max(round(info.video.frame_rate), 1)
    gop_size = fps * KEYFRAME_INTERVAL_SECS
    return (
        "ffmpeg",
        "-hide_banner",
        "-y",
        "-i",
        str(info.path),
        # Video encoding
        "-c:v",
        "libx264",
        "-preset",
        PRESET,
        "-crf",
        str(CRF),
        "-profile:v",
        PROFILE,
        "-level:v",
        LEVEL,
        "-pix_fmt",
        "yuv420p",
        # GOP structure
        "-g",
        str(gop_size),
        "-keyint_min",
        str(fps),
        "-bf",
        str(B_FRAMES),
        # x264 tuning
        "-x264-params",
        X264_PARAMS,
        # Audio: stream copy
        "-c:a",
        "copy",
        # Stream selection
        "-map",
        "0:v:0",
        "-map",
        "0:a:0",
        # Container optimisation
        "-movflags",
        "+faststart",
        "-map_metadata",
        "-1",
        # Progress output on stdout
        "-progress",
        "pipe:1",
        str(output_path),
    )


def build_remux_cmd(
    info: FileInfo,
    output_path: Path,
    /,
) -> tuple[str, ...]:
    """
    Build the ffmpeg command for container-level remux optimisation.

    Stream-copies all codecs, selects first video and audio streams,
    applies faststart, and strips metadata and chapters.
    """
    return (
        "ffmpeg",
        "-hide_banner",
        "-y",
        "-i",
        str(info.path),
        # Stream copy (no re-encoding)
        "-c:v",
        "copy",
        "-c:a",
        "copy",
        # Stream selection
        "-map",
        "0:v:0",
        "-map",
        "0:a:0",
        # Container optimisation
        "-movflags",
        "+faststart",
        "-map_metadata",
        "-1",
        "-map_chapters",
        "-1",
        # Progress output on stdout
        "-progress",
        "pipe:1",
        str(output_path),
    )


def build_webm_remux_cmd(
    info: FileInfo,
    output_path: Path,
    /,
) -> tuple[str, ...]:
    """
    Build the ffmpeg command for MKV-to-WebM remux optimisation.

    Stream-copies AV1 video and Opus audio, selects first video and audio
    streams, places Cues (seek index) at front, aligns clusters to
    2-second keyframe boundaries, and strips metadata and chapters.
    """
    return (
        "ffmpeg",
        "-hide_banner",
        "-y",
        "-i",
        str(info.path),
        # Stream copy (no re-encoding)
        "-c:v",
        "copy",
        "-c:a",
        "copy",
        # Stream selection
        "-map",
        "0:v:0",
        "-map",
        "0:a:0",
        # WebM container format
        "-f",
        "webm",
        # Container optimisation
        "-cues_to_front",
        "true",
        "-cluster_time_limit",
        "2000",
        "-write_crc32",
        "false",
        # Strip metadata
        "-map_metadata",
        "-1",
        "-map_chapters",
        "-1",
        # Progress output on stdout
        "-progress",
        "pipe:1",
        str(output_path),
    )

File: src/WebOptimise/Commands.fs
namespace WebOptimise

[<Struct; RequireQualifiedAccess>]
type VideoAction =
    | Copy
    | Encode of
        preset: string *
        crf: int *
        profile: string *
        level: string *
        gopSize: int *
        minKeyint: int *
        bframes: int *
        x264Params: string

[<Struct; RequireQualifiedAccess>]
type AudioAction = | Copy

[<Struct; RequireQualifiedAccess>]
type Container =
    | Mp4Faststart
    | WebmDash of clusterTimeLimit: int

[<NoComparison; NoEquality>]
type FfmpegCmd = {
    Input: MediaFilePath
    Output: OutputPath
    Video: VideoAction
    Audio: AudioAction
    Container: Container
    StripMetadata: bool
    StripChapters: bool
}

/// Pure ffmpeg command builders. No I/O.
[<RequireQualifiedAccess>]
module Commands =

    let render (cmd: FfmpegCmd) : string list = [
        "-hide_banner"
        "-y"
        "-i"
        MediaFilePath.value cmd.Input
        yield!
            match cmd.Video with
            | VideoAction.Copy -> [
                "-c:v"
                "copy"
              ]
            | VideoAction.Encode(preset, crf, profile, level, gopSize, minKeyint, bframes, x264Params) ->
                [
                    "-c:v"
                    "libx264"
                    "-preset"
                    preset
                    "-crf"
                    crf.ToString()
                    "-profile:v"
                    profile
                    "-level:v"
                    level
                    "-pix_fmt"
                    "yuv420p"
                    "-g"
                    gopSize.ToString()
                    "-keyint_min"
                    minKeyint.ToString()
                    "-bf"
                    bframes.ToString()
                    "-x264-params"
                    x264Params
                ]
        "-c:a"
        "copy"
        "-map"
        "0:v:0"
        "-map"
        "0:a:0"
        yield!
            match cmd.Container with
            | Container.Mp4Faststart -> [
                "-movflags"
                "+faststart"
              ]
            | Container.WebmDash clusterTimeLimit ->
                [
                    "-f"
                    "webm"
                    "-cues_to_front"
                    "true"
                    "-cluster_time_limit"
                    clusterTimeLimit.ToString()
                    "-write_crc32"
                    "false"
                ]
        "-map_metadata"
        "-1"
        if cmd.StripChapters then
            "-map_chapters"
            "-1"
        "-progress"
        "pipe:1"
        OutputPath.value cmd.Output
    ]

    let buildEncode (info: MediaFileInfo) (output: OutputPath) : FfmpegCmd =
        let fps = max (int (round info.Video.FrameRate)) 1
        let gopSize = fps * Constants.KeyframeIntervalSecs

        {
            Input = info.Path
            Output = output
            Video =
                VideoAction.Encode(
                    Constants.Preset,
                    Constants.Crf,
                    Constants.Profile,
                    Constants.Level,
                    gopSize,
                    fps,
                    Constants.BFrames,
                    Constants.X264Params
                )
            Audio = AudioAction.Copy
            Container = Container.Mp4Faststart
            StripMetadata = true
            StripChapters = false
        }

    let buildRemux (info: MediaFileInfo) (output: OutputPath) : FfmpegCmd = {
        Input = info.Path
        Output = output
        Video = VideoAction.Copy
        Audio = AudioAction.Copy
        Container = Container.Mp4Faststart
        StripMetadata = true
        StripChapters = true
    }

    let buildWebmRemux (info: MediaFileInfo) (output: OutputPath) : FfmpegCmd = {
        Input = info.Path
        Output = output
        Video = VideoAction.Copy
        Audio = AudioAction.Copy
        Container = Container.WebmDash 2000
        StripMetadata = true
        StripChapters = true
    }
